File: HackerRank/WeekOfCode/WeekOfCode32/GeometricTrick.cs
namespace HackerRank.WeekOfCode32.GeometricTrick
{
	// https://www.hackerrank.com/contests/w32/challenges/geometric-trick

	using System;
	using System.Collections;
	using System.Collections.Generic;
	using System.Diagnostics;
	using System.IO;
	using System.Linq;
	using System.Text;
	using static FastIO;

	public class Solution
	{
		public void solve(Stream input, Stream output)
		{
			InitInput(input);
			InitOutput(output);
			solve();
			Flush();
		}

		int n;
		char[] s;
		int[] factors;

		public void solve()
		{
			n = Ni();
			s = Nc(n);

			factors = PrimeFactorsUpTo(n);

			long count = 0;
			int j1;
			for (int j = 0; j < n; j++)
			{
				if (s[j] != 'b') continue;
				j1 = j + 1;
				EnumerateFactorsX2(factors, j1, j1, f =>
				{
					if (s[f - 1] == 'b') return;
					var f2 = 1L * j1 * j1 / f;
					Console.Error.WriteLine($"{j1},{f},{f2}");
					if (f2 > n || s[f - 1] != (s[f2 - 1] ^ 2)) return;
					count++;
				});
			}

			WriteLine(count);
		}

		public static int[] PrimeFactorsUpTo(int n)
		{
			var factors = new int[n + 1];

			for (int i = 2; i <= n; i += 2)
				factors[i] = 2;

			var sqrt = (int)Math.Sqrt(n);
			for (int i = 3; i <= sqrt; i += 2)
			{
				if (factors[i] != 0) continue;
				for (int j = i * i; j <= n; j += i + i)
				{
					if (factors[j] == 0)
						factors[j] = i;
				}
			}

			for (int i = 3; i <= n; i += 2)
			{
				if (factors[i] == 0)
					factors[i] = i;
			}

			return factors;
		}

		public static int EnumerateFactorsX2(int[] factors, long n, int max, Action<long> action = null, long f = 1)
		{
			if (f > max)
				return 0;

			if (n == 1)
			{
				action?.Invoke(f);
				return 1;
			}

			long p = factors[n];
			long c = 1;
			long next = n / p;
			while (next > 1 && factors[next] == p)
			{
				c++;
				next = next / p;
			}

			c *= 2;
			int result = EnumerateFactorsX2(factors, next, max, action, f);
			while (c-- > 0)
			{
				f *= p;
				result += EnumerateFactorsX2(factors, next, max, action, f);
			}

			return result;
		}

	}


}

File: HackerRank/WeekOfCode/WeekOfCode32/GeometricTrick.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {

    static long geometricTrick(String str){
        s = str;
        n = str.length();
        
        factors = new int[n+1];
        for (int i=0; i<=n; i++)
            factors[i] = i%2==0 ? 2 : i;

        int sq = (int) Math.sqrt(n);
        for (int i=3; i<=sq; i++)
            if (factors[i] == i)
            {
                for (int j=i*i; j<=n; j+=2*i)
                    if (factors[j] == j) 
                        factors[j] = i;
            }

        long count = 0;
        for (int j=0; j<n; j++)
            if (s.charAt(j)=='b')
                count += dfs(j+1, j+1, 1);
        return count;
    }
    
    static int n;
    static String s;
    static int factors[];

    static long dfs(int d, int j, long f)
    {
        if (f > j) return 0;
        if (d==1)
        {
            int ff = (int)f;
            if (s.charAt(ff-1)=='b') return 0;
            long f2 = ((long)j*j/f);
            if (f2<=n && (int)s.charAt(ff-1)==((int)s.charAt((int)f2-1)^2)) return 1;
            return 0;
        }

        int p = factors[d];
        int c = 1;
        int next = d/p;
        while (next > 1 && factors[next]==p) { c++; next/=p; }

        c*=2;
        long res = dfs(next, j, f);
        while (c-- >0)
        {
            f*=p;
            res += dfs(next, j, f);
        }
        return res;        
    }


    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        String s = in.next();
        long result = geometricTrick(s);
        System.out.println(result);
    }
}
